crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4210,6 +4441,46 @@ impl<'db> UnionType<'db> {
         Self::from_elements(db, self.elements(db).iter().map(transform_fn))
     }
 
+    pub(crate) fn map_with_boundness(
```

Comment:
This whole logic here was simply extracted into a function. The logic itself didn't change.